docs/zkapps/tutorials/06-offchain-storage.mdx

```diff
@@ -0,0 +1,390 @@
+
+### Offchain Storage and Decentralization
+
+This solution works well as long as the server is being run as expected. This is reasonable when the users of the application and the developer both trust the server the zkApp is using.
+
+While this is useful for prototyping applications, and even putting many applications into production, its limited from a decentralization standpoint.
+
+In the short-run, we expect the following solutions for data availability:
+
+* Single-server offchain storage (the solution presented here).
+* Multi-server offchain storage - with guarantees as long as a majority of servers behave honestly (Grant available; see below)
+* Modular blockchain offchain storage (Several dependencies, but under exploration)
+
+In the long run, a feature is being considered for a future hard fork to add purchasable data storage with data availability guarantees to Mina itself. However, even then, putting more data on chain burdens the network, so it is likely that putting large amounts of data on chain will be relatively expensive. Given that it is expected that the short-run solutions presented above will still be relevant and useful.
+
+### Further development
+
+We're looking for a developer to run with the library presented here, improve it, and run instances for the community. There is a grant opportunity for taking this on. If interested, please reach out [here](mailto:build@minaprotocol.com).
+
+Some suggested improvements:
+
+* Eliminate DDOS vulnerability by adding a token that limits storage requests
+* Do not store trees for a smart contract if that contract is misconfigured in a way to prevent cleaning up old data.
+* Add support to the client library for connecting to multiple storage servers, enabling correctness under majority-honest.
+* Switch the project to a more scalable database implementation (eg Redis)
+* Write an implementation for an automatically scalable service (eg Cloudflare)
+
+## Implementing a Project Using Offchain Storage
+
+As usual, there is sample code for this project, which you can find [here](https://github.com/es92/zkApp-examples/tree/main/06-offchain-storage/contracts).
+
+We will be writing and discussing the [src/main.ts](https://github.com/es92/zkApp-examples/blob/main/06-offchain-storage/contracts/src/main.ts) and [src/NumberTreeContract.ts](https://github.com/es92/zkApp-examples/blob/main/06-offchain-storage/contracts/src/NumberTreeContract.ts) files in this project.
+
+This project implements a tree, where each leaf is either empty or stores a number. Updates to the tree can update a leaf, if the new number in the leaf is greater than the old number. The root of the tree is stored on chain, while the tree itself is stored on an offchain storage server.
```

Comment:
`This project implements a tree, where each leaf is either empty or stores a number` > would append `, which will be our data.` (If that's correct, I haven't gotten further editing yet.)

`off-chain` (hypen)